File: 119-Pascal's Triangle II/pascal_triangle_2.dart
main() {
  final rowIndex = 4;
  PascalTriangle2.getRow(rowIndex);
}

abstract class PascalTriangle2 {
  static List<int> getRow(int rowIndex) {
    //
    // First attempt 26 June 2024 13:09
    // Runtime: 330ms, Memory: 146.70MB
    //
    // final res = [1];
    // var prev = 1;
    //
    // for (int i = 1; i <= rowIndex; i++) {
    //   final curr = prev.toDouble() * (rowIndex - i + 1) ~/ i;
    //   res.add(curr);
    //   prev = curr;
    // }
    //
    // return res;
    //
    // Second attempt 26 June 2024
    // Runtime: 300ms, Memory: 151.76MB
    // Recursive
    //
    if (rowIndex == 0) return [1];

    final prevRow = getRow(rowIndex - 1);
    final row = <int>[];

    row.add(1);
    for (int i = 1; i < rowIndex; i++) {
      row.add(prevRow[i - 1] + prevRow[i]);
    }
    row.add(1);

    return row;
  }
}

File: 119-Pascal's Triangle II/PascalTriangle2.kt
package main

fun main() {
    val rowIndex = 4
    println(PascalTriangle2.getRow(rowIndex))
}

object PascalTriangle2 {
    fun getRow(rowIndex: Int): List<Int> {
        //
        // First attempt 26 June 2024 12:59
        // Runtime: 143ms, Memory: 34.76MB
        //
//        val res = mutableListOf(1)
//        var prev = 1
//
//        for (i in 1..rowIndex) {
//            val curr = (prev.toLong() * (rowIndex - i + 1) / i).toInt()
//            res.add(curr)
//            prev = curr
//        }
//
//        return res
        //
        // Second attempt 26 June 2024 13:03
        // Runtime: 140ms, Memory: 34.20MB
        // Recursive
        //
        if (rowIndex == 0) return listOf(1)

        val prevRow = getRow(rowIndex - 1)
        val row = mutableListOf<Int>()

        row.add(1)
        for (i in 1..<rowIndex) {
            row.add(prevRow[i] + prevRow[i - 1])
        }
        row.add(1)

        return row
    }
}

